pkg/engine2/edge_targets.go

```diff
@@ -0,0 +1,297 @@
+		if err != nil {
+			return false, err
+		}
+		paths, err := graph.AllPathsBetween(pathSelectionGraph, edge.Source, edge.Target)
+		if len(paths) == 0 || err != nil {
+			return false, err
+		}
+
+		hasValidPath := false
+		for _, path := range paths {
+			if path_selection.PathSatisfiesClassification(e.Kb, path, classification) {
+				hasValidPath = true
+				break
+			}
+		}
+		if !hasValidPath {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func ReadGetValidEdgeTargetsConfig(path string) (GetValidEdgeTargetsConfig, error) {
+	var config GetValidEdgeTargetsConfig
+	yamlFile, err := os.ReadFile(path)
+	if err != nil {
+		return config, err
+	}
+	err = yaml.Unmarshal(yamlFile, &config)
+	if err != nil {
+		return config, err
+	}
+	return config, nil
+}
+
+/*
+GetValidEdgeTargets returns a map of valid edge targets for each source resource in the supplied graph.
+The returned map is keyed by the source resource's string representation.
+The value for each source resource is a list of valid target resources.
+
+Targets are considered valid if there is a set of kb paths between the source and target
+that satisfies both source and target path satisfaction classifications.
+
+A partial set of valid targets can be generated using the filter criteria in the context's config.
+*/
+func (e *Engine) GetValidEdgeTargets(context *GetPossibleEdgesContext) (map[string][]string, error) {
+
+	inputGraph, err := unmarshallInputGraph(context.InputGraph)
+	if err != nil {
+		return nil, err
+	}
+	solutionCtx := NewSolutionContext(e.Kb)
+	err = solutionCtx.LoadGraph(inputGraph)
+	if err != nil {
+		return nil, err
+	}
+	topologyGraph, err := e.GetViewsDag(DataflowView, solutionCtx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var sources []construct.ResourceId
+	var targets []construct.ResourceId
+
+	qualifiedTypeMatcher := func(id construct.ResourceId) func(otherType construct.ResourceId) bool {
+		return func(otherType construct.ResourceId) bool {
+			return otherType.QualifiedTypeName() == id.QualifiedTypeName()
+		}
+	}
+
+	// filter resources based on the context
+	err = construct.WalkGraph(topologyGraph, func(id construct.ResourceId, resource *construct.Resource, nerr error) error {
+		tag := e.GetResourceVizTag(string(DataflowView), id)
+		if len(context.Tags) > 0 && !slices.Contains(context.Tags, tag) {
+			return nil
+		}
+		isSource := true
+		isTarget := true
+		if len(context.Resources.Sources) > 0 && !slices.Contains(context.Resources.Sources, id) {
+			isSource = false
+		}
+		if len(context.Resources.Targets) > 0 && !slices.Contains(context.Resources.Targets, id) {
+			isTarget = false
+		}
+
+		if len(context.ResourceTypes.Sources) > 0 && !slices.ContainsFunc(context.ResourceTypes.Sources, qualifiedTypeMatcher(id)) {
+			isSource = false
+		}
+
+		if len(context.ResourceTypes.Targets) > 0 && !slices.ContainsFunc(context.ResourceTypes.Targets, qualifiedTypeMatcher(id)) {
+			isTarget = false
+		}
```

Comment:
why have this check when you check for if the sources or targets contain the id before hand?